runtime/service.go

```diff
@@ -436,10 +345,266 @@ func (s *service) nextVSockPort() uint32 {
 	return port
 }
 
+func (s *service) waitVMReady() error {
+	select {
+	case <-s.vmReady:
+		return nil
+	case <-time.After(firecrackerStartTimeout):
+		return status.Error(codes.DeadlineExceeded, "timed out waiting for VM start")
+	}
+}
+
+// CreateVM will attempt to create the VM as specified in the provided request, but only on the first request
+// received. Any subsequent requests will be ignored and get an AlreadyExists error response.
+func (s *service) CreateVM(requestCtx context.Context, request *proto.CreateVMRequest) (*empty.Empty, error) {
+	defer logPanicAndDie(s.logger)
+
+	errCh := make(chan error, 1)
+	s.vmStartOnce.Do(func() {
+		errCh <- s.createVM(requestCtx, request)
+		close(errCh)
+	})
+
+	select {
+	case err := <-errCh:
```

Comment:
I think it might be simpler to structure this without a channel, since we're not really doing any cross-goroutine communication or concurrency.  What about something like this?

```go
calledCreate := false
var err error
s.vmStartOnce.Do(func() {
	err = s.createVM(requestCtx, request)
	calledCreate = true
}

if !calledCreate {
	return nil, status.Error(codes.AlreadyExists, "shim cannot create VM more than once")
}

// everything else
```

Reply:
I had that originally but liked the idea of unifying `err` and `calledCreate` with a channel, which is why I went with what I have. It just seems nice that the thing that determines if you ran will also tell you what, if anything, went wrong during the run.

This is splitting hairs in my mind though; is the current solution actually confusing enough to warrant updating? Happy to update if so.

Reply:
The use of a channel caused me to spend some time wondering about whether the `case err := <-errCh:` would be hit reliably, so it took me a bit more time to read.  There's also a bit of a cosmetic thing about reducing indentation.  But there's nothing wrong about this code, these are just stylistic.  I prefer the way I suggested, but not to such a huge amount.

> It just seems nice that the thing that determines if you ran will also tell you what, if anything, went wrong during the run.

Yeah, I really wish Go had a `Result<Ok<V>, Err<E>>`.

Reply:
Okay that makes sense, I'll update then, no need to make things unnecessarily confusing when all else is equal.

And hopefully Go generics will get us `Result` someday...